### PR TITLE
Week4 blob

### PR DESCRIPTION
**Pull request by Xin128**
Finally import blobstore with Username URL
Well-tested features:
1) combination of datastore and blobstore
2) display photos fetched from blobstore service
3) hashmap data structure for saving data of a user 

[Question]: Maybe a user class is more necessary based on this scenario?

### CODE REVIEW

File: portfolio/pom.xml [hunk 1]
```diff
@@ -44,7 +44,7 @@
         <artifactId>appengine-maven-plugin</artifactId>
         <version>2.2.0</version>
         <configuration>
-          <deploy.projectId>xinhaofighting-step-2020</deploy.projectId>
+          <deploy.projectId>google.com:xinhaofighting-step</deploy.projectId>
```
Comment (cisxingg) [inline portfolio/pom.xml]: You may not need the "google.com:", also maybe we don't even need to change this if the new one is not working :(

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -39,17 +42,31 @@
...
   protected static final String COMMENT = "Comment";
-  private final String CONTENT_PROPERTY = "content";
+  protected static final String CONTENT_PROPERTY = "content";
   private final int DEFAULT_MAX_COMMENT_NUM = 1;
+  private final String DEFAULT_USERNAME = "defaultUser";
   private final String INPUT_MSG_FORM = "comment-input";
   private final String NUM_COMMENT_FORM = "numComment";
-  private final String TIMESTAMP_PROPERTY = "timestamp";
+  protected static final String TIMESTAMP_PROPERTY = "timestamp";
   private final String USEREMAIL_PROPERTY = "userEmail";
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: order the variables for https://google.github.io/styleguide/javaguide.html#s4.8.7-modifiers

File: portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    PrintWriter out = response.getWriter();
+    out.println("<p>Here's the image you uploaded:</p>");
+    out.println("<a href=\"" + imageUrl + "\">");
+    out.println("<img src=\"" + imageUrl + "\" />");
+    out.println("</a>");
+    out.println("<p>Here's the text you entered:</p>");
+    out.println(message);    
+    
+    // // Redirect back to the current page
+    // response.sendRedirect("/index.html");
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java]: Do you need this? Don't keep the commented-out code in the code base :)

File: portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    commentEntity.setProperty(CONTENT_PROPERTY, message);
+    commentEntity.setProperty(TIMESTAMP_PROPERTY,timestamp);
+    commentEntity.setProperty(BLOB_URL_PROPERTY, imageUrl);
+
+    // Used Datastore survice to store newly created comment entity
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(commentEntity);
+
+    // Output some HTML that shows the data the user entered.
+    // A real codebase would probably store these in Datastore.
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java]: What does this mean, a real code base? You are making a real database :D

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -39,17 +42,31 @@
...
 
 
+
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String requestedBlobComment = request.getParameter(BLOB_URL);
+    if (requestedBlobComment != null) {
+        BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: 2 spaces.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -62,16 +79,24 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
 
-    // Add limited number of comment contents and user email to the userComment Map  
-    HashMap<String, ArrayList<String>> userComment = new HashMap<String, ArrayList<String>>();
+    /* Add limited number of comment contents and user email to the userComment Map  
+     * Note: commeentMap data structure: 
+     * {userEmail: [[userComment1, userCommentImage1],[userComment2, userCommentImage2]...]}
+     */
+    HashMap<String, ArrayList<ArrayList<String>>> userComment = new HashMap<String, ArrayList<ArrayList<String>>>();
     for (Entity commentEntity : results.asList(FetchOptions.Builder.withLimit(maxNumComments))) {
         String commentMsg = (String) commentEntity.getProperty(CONTENT_PROPERTY);
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Two spaces inside for loop.

File: portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+ */
+@WebServlet("/my-form-handler")
+public class FormHandlerServlet extends HttpServlet {
+  String BLOB_URL_PROPERTY = DataServlet.BLOB_URL_PROPERTY;
+  String COMMENT = DataServlet.COMMENT;
+  String CONTENT_PROPERTY = DataServlet.CONTENT_PROPERTY;
+  String TIMESTAMP_PROPERTY = DataServlet.TIMESTAMP_PROPERTY;
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java]: Delete extra line.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 8]
```diff
@@ -80,26 +105,6 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: This method does nothing now? Do we still need it?

File: portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    // We could check the validity of the file here, e.g. to make sure it's an image file
+    // https://stackoverflow.com/q/10779564/873165
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java]: Add todo if you plan to do this later:
TODO: We could...

File: portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+    // Output some HTML that shows the data the user entered.
+    // A real codebase would probably store these in Datastore.
+    PrintWriter out = response.getWriter();
+    out.println("<p>Here's the image you uploaded:</p>");
+    out.println("<a href=\"" + imageUrl + "\">");
+    out.println("<img src=\"" + imageUrl + "\" />");
+    out.println("</a>");
+    out.println("<p>Here's the text you entered:</p>");
+    out.println(message);
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java]: Question: Do you need this?

File: portfolio/pom.xml (hunk 1 above)
Comment (Xin128) [reply portfolio/pom.xml]: :( I will keep the "xinhaofighting-step" for now... Let's see if we need to switch back later then~

File: portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java (hunk 4 above)
Comment (Xin128) [reply portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java]: Explanations associated with the code below: 
This is just a milestone from the blobstore step, which seems unnecessary since we have done the rest. I just realize that this whole displaying page can be removed since we fetch that back to our original index.html. 
Conclusion: 
I have removed comment and all unnecessary code.

Comment (Xin128) [reply portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java]: After testing, I added the redirecting feature to ensure user will not go to that uploading page for image storage

File: portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java (hunk 9 above)
Comment (Xin128) [reply portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java]: Already checked the validity in dataservelet, no need for the TODO

**Review by warfa-google — COMMENTED** [review]
This will be refactored as discussed in our 1:1

File: portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java [hunk 11]
```diff
@@ -43,13 +42,12 @@
...
 @WebServlet("/my-form-handler")
 public class FormHandlerServlet extends HttpServlet {
-  String BLOB_URL_PROPERTY = DataServlet.BLOB_URL_PROPERTY;
-  String COMMENT = DataServlet.COMMENT;
-  String CONTENT_PROPERTY = DataServlet.CONTENT_PROPERTY;
-  String TIMESTAMP_PROPERTY = DataServlet.TIMESTAMP_PROPERTY;
+  private String BLOB_URL_PROPERTY = DataServlet.BLOB_URL_PROPERTY;
+  private String COMMENT = DataServlet.COMMENT;
+  private String CONTENT_PROPERTY = DataServlet.CONTENT_PROPERTY;
+  private String TIMESTAMP_PROPERTY = DataServlet.TIMESTAMP_PROPERTY;
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java:48]: We don't need to make all these extra variables, you can simply call DataServlet.BLOB_URL_PROPERTY in the code below.

File: portfolio/src/main/java/com/google/sps/servlets/UserLoginServelet.java [hunk 12]
```diff
@@ -14,29 +14,24 @@
...
-import java.util.HashMap;
 
 /**
   * Servlet that ensure User's log in status.
   * Note: We first added the Gson library dependency to pom.xml.
   */
 @WebServlet("/userLogIn")
 public class UserLoginServelet extends HttpServlet {
-  HashMap<String, Object> userInfo = new HashMap<String, Object>();
   final private String redirectStr = "/";
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/UserLoginServelet.java]: Naming for a final value: https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 13]
```diff
@@ -83,30 +68,25 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
      * Note: commeentMap data structure:
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: spelling error.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 14]
```diff
@@ -83,30 +68,25 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+      // Get the different properties (userEmail, message, image) of a comment Entity
+      String commentMsg = (String) commentEntity.getProperty(CONTENT_PROPERTY);
+      String commentUser = (String) commentEntity.getProperty(USEREMAIL_PROPERTY);
+      if (commentUser == null) {
+        commentUser = DEFAULT_USERNAME;
+      }
+      String imageUrl = (String) commentEntity.getProperty(BLOB_URL_PROPERTY);
+      ArrayList<ArrayList<String>> msglist = userComment.get(commentUser);
+      if (msglist == null) {
+        ArrayList<String> firstComment = new ArrayList<String>(Arrays.asList(commentMsg,imageUrl));
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: space in between "," and "imageUrl". Same below.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 15]
```diff
@@ -83,30 +68,25 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
      * Note: commeentMap data structure: 
      * {userEmail: [[userComment1, userCommentImage1],[userComment2, userCommentImage2]...]}
      */
-    HashMap<String, ArrayList<ArrayList<String>>> userComment = new HashMap<String, ArrayList<ArrayList<String>>>();
+    HashMap<String, ArrayList<ArrayList<String>>> userComment = new HashMap();
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I don't really like the idea of ArrayList\<ArrayList\>>when you know the size of the inside ArrayList is fixed. I think a class called UserComment to contain both fields is preferred here. If it is hard to parse it to json, maybe we can try java tuple or some equivalent data structures?

Comment (warfa-google) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: agreed. A class to encapsulate the comment and picture would look better.

File: portfolio/src/main/webapp/script.js [hunk 16]
```diff
@@ -35,19 +36,52 @@ function addRandomDescription() {
...
+// delete all the comments from Datastore
 function deleteAllComments() {
   const params = new URLSearchParams();
   fetch("/delete-data", { method: "POST", body: params });
 }
 
+// fetch the url of image blobstore and show the form to upload files
+function fetchBlobstoreUrlAndShowForm() {
+  const messageForm = document.getElementById('my-form');
+  var url = "/data?blob_upload_url=" + messageForm;
```
Comment (warfa-google) [inline portfolio/src/main/webapp/script.js]: doesn't this need to be changed to /blob?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 17]
```diff
@@ -58,17 +56,17 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
 
-    /* Add limited number of comment contents and user email to the userComment Map  
-     * Note: commeentMap data structure: 
-     * {userEmail: [[userComment1, userCommentImage1],[userComment2, userCommentImage2]...]}
-     */
-    HashMap<String, ArrayList<ArrayList<String>>> userComment = new HashMap();
+    /* Add limited number of comment contents and user email to the userComment Map
+     * Note: commentMap data structure:
+     * {UserName1: UserComment1, UserName2: UserComment2, UserName3: UserComment3...}
+     *      */
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:67]: nit: comment looks off.

File: portfolio/src/main/java/com/google/sps/servlets/UserComment.java [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import org.javatuples.Pair;
+
+public class UserComment {
+  // wrapped comment entity that contains the comment message and image url
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/UserComment.java]: nit: Capitalize "W"

File: portfolio/src/main/java/com/google/sps/servlets/UserComment.java [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+  private Pair<String, String> commentEntity;
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/UserComment.java]: Why pair if you have a separate class already?
Can simply have:
private String commentMsg;
private String imageUrl;

File: portfolio/src/main/webapp/script.js [hunk 20]
```diff
@@ -35,19 +36,53 @@ function addRandomDescription() {
...
-}
 
+  /* fetch from data url, then fetch from remote blobstore url with image and display it on the webpage
+   * Note: commentMap data structure:
+   * {UserName1: [User1Comment1, User1Comment2...],
+   * UserName2: [User2Comment1, User2Comment2...],...}
+   */
+  var commentContainer = document.getElementById('comment-container');
+  fetch(url).then(response => response.json()).then((commentMap) => {
+    Object.entries(commentMap).forEach(([commentUser,commentEntity]) => {
```
Comment (cisxingg) [inline portfolio/src/main/webapp/script.js:51]: nit: space between "," and "commentEntity"

File: walkthroughs/week-5-tdd/project/gcloud-tutorials-calendar.iml [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+    </content>
+    <orderEntry type="inheritedJdk" />
+    <orderEntry type="sourceFolder" forTests="false" />
+    <orderEntry type="library" scope="PROVIDED" name="Maven: javax.servlet:javax.servlet-api:4.0.1" level="project" />
+    <orderEntry type="library" name="Maven: com.google.code.gson:gson:2.8.6" level="project" />
+    <orderEntry type="library" scope="TEST" name="Maven: junit:junit:4.12" level="project" />
+    <orderEntry type="library" scope="TEST" name="Maven: org.hamcrest:hamcrest-core:1.3" level="project" />
+    <orderEntry type="library" name="Maven: asm:asm:3.3.1" level="project" />
+  </component>
+</module>
```
Comment (cisxingg) [inline walkthroughs/week-5-tdd/project/gcloud-tutorials-calendar.iml]: Is this file meant to be in the PR?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 22]
```diff
@@ -53,55 +52,40 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
     Query query = new Query(COMMENT).addSort(TIMESTAMP_PROPERTY, SortDirection.DESCENDING);
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     PreparedQuery results = datastore.prepare(query);
 
-    // Add limited number of comment contents and user email to the userComment Map  
-    HashMap<String, ArrayList<String>> userComment = new HashMap<String, ArrayList<String>>();
+    /* Add limited number of comment entity  and user email to the userList Map
+     * Note: userList data structure:
+     * {UserName1: [User1Comment1, User1Comment2...],
+     * UserName2: [User2Comment1, User2Comment2...],...}
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: This may need to be updated.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 23]
```diff
@@ -53,55 +52,40 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+      if (commentUser == null) {
+        commentUser = DEFAULT_USERNAME;
+      }
+      String imageUrl = (String) commentEntity.getProperty(BLOB_URL_PROPERTY);
+      ArrayList<UserComment> msglist = userList.get(commentUser);
+      UserComment newComment = new UserComment(commentUser, commentMsg, imageUrl);
+      if (msglist == null) {
+        userList.put(commentUser, new ArrayList<>(Arrays.asList(newComment)));
+      } else {
+        msglist.add(newComment);
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:82]: Do you need to update userList?

Comment (cisxingg) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:82]: This is a wrong comment, I will delete it.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 24]
```diff
@@ -53,55 +52,39 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
     // Create the query and prepared query to load comment entities from database
     Query query = new Query(COMMENT).addSort(TIMESTAMP_PROPERTY, SortDirection.DESCENDING);
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     PreparedQuery results = datastore.prepare(query);
 
-    // Add limited number of comment contents and user email to the userComment Map  
-    HashMap<String, ArrayList<String>> userComment = new HashMap<String, ArrayList<String>>();
+    /* Add limited number of comment entity and user email to the userList Map
+     * Note: userList data structure:
+     * {User1: ArrayList<UserComment>, User2: ArrayList<UserComment>,...}
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: space in between "{" and "User1".

File: portfolio/src/main/java/com/google/sps/servlets/UserLoginServelet.java [hunk 25]
```diff
@@ -14,42 +14,37 @@
...
 
 /**
   * Servlet that ensure User's log in status.
   * Note: We first added the Gson library dependency to pom.xml.
   */
 @WebServlet("/userLogIn")
 public class UserLoginServelet extends HttpServlet {
-  HashMap<String, Object> userInfo = new HashMap<String, Object>();
-  final private String redirectStr = "/";
+  final private String REDIRECTSTR = "/";
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/UserLoginServelet.java]: nit: REDIRECT_STR

File: portfolio/src/main/java/com/google/sps/servlets/UserComment.java [hunk 26]
```diff
@@ -0,0 +1,50 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import org.javatuples.Pair;
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/UserComment.java]: Delete?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 27]
```diff
@@ -53,55 +52,39 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+      String imageUrl = (String) commentEntity.getProperty(BLOB_URL_PROPERTY);
+      ArrayList<UserComment> msglist = userList.get(commentUser);
+      UserComment newComment = new UserComment(commentUser, commentMsg, imageUrl);
+      if (msglist == null) {
+        userList.put(commentUser, new ArrayList<>(Arrays.asList(newComment)));
+      } else {
+        msglist.add(newComment);
+      }
     }
+    System.out.println(new Gson().toJson(userList));
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Delete?

File: portfolio/.bash_profile [hunk 28]
```diff
@@ -0,0 +1,2 @@
+
```
Comment (cisxingg) [inline portfolio/.bash_profile]: Extra line?
Do you need to update this file?

**Review by cisxingg — APPROVED** [review]
Nits, but otherwise LGTM.

File: portfolio/src/main/java/com/google/sps/servlets/BlobServlet.java [hunk 29]
```diff
@@ -0,0 +1,25 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Servlet that returns images in blobstore.
+ * Note: We first added the Gson library dependency to pom.xml.
+ */
+@WebServlet("/blob")
+public class BlobServlet extends HttpServlet {
+  // Process the Blob URL
```
Comment (warfa-google) [inline portfolio/src/main/java/com/google/sps/servlets/BlobServlet.java:17]: this comment is unnecessary.

File: portfolio/src/main/webapp/script.js [hunk 30]
```diff
@@ -35,19 +36,55 @@ function addRandomDescription() {
 }
 
 function getCommentInForm() {
+  // get number of limited comments from input form
```
Comment (warfa-google) [inline portfolio/src/main/webapp/script.js:39]: remove this comment

File: portfolio/src/main/webapp/script.js [hunk 31]
```diff
@@ -35,19 +36,55 @@ function addRandomDescription() {
...
+          imgElem.src = objectURL;
+          commentContainer.append(imgElem);
+          commentContainer.append(linebreak);
+        });
+      });
+    })
+  }); 
+};
+
+// delete all the comments from Datastore
```
Comment (warfa-google) [inline portfolio/src/main/webapp/script.js:72]: unnecessary comment

File: portfolio/src/main/webapp/script.js [hunk 32]
```diff
@@ -35,19 +36,55 @@ function addRandomDescription() {
...
 function deleteAllComments() {
   const params = new URLSearchParams();
   fetch("/delete-data", { method: "POST", body: params });
 }
 
+// fetch the url of image blobstore and show the form to upload files
```
Comment (warfa-google) [inline portfolio/src/main/webapp/script.js:78]: given the name of the function this is comment is unnecessary.

**Review by warfa-google — APPROVED** [review]
Looks good to me in general. There are a few obvious comments you should remove. In general try to strike a balance between commenting things that are complicated and commenting every line. A good rule of thumb is not to comment the What or How, as that is usually self-explanatory from the code, but instead comment the Why as that is hard to intuit by reading the code.

**Comment by cisxingg** [comment]
Submit at some point? :D
